clippy_lints/src/matches/redundant_guard.rs

```diff
@@ -0,0 +1,226 @@
+        else if let Guard::If(if_expr) = guard
+            && let ExprKind::Binary(bin_op, local, pat) = if_expr.kind
+            && matches!(bin_op.node, BinOpKind::Eq)
+            && let Some((pat_binding, field_name)) = get_pat_binding(cx, local, outer_arm)
+            && expr_can_be_pat(cx, pat)
+            // Ensure they have the same type. If they don't, we'd need deref coercion which isn't
+            // possible (currently) in a pattern. In some cases, you can use something like
+            // `as_deref` or similar but in general, we shouldn't lint this as it'd create an
+            // extraordinary amount of FPs.
+            //
+            // This isn't necessary in the other two checks, as they must be a pattern already.
+            && cx.typeck_results().expr_ty(local) == cx.typeck_results().expr_ty(pat)
+        {
+            emit_redundant_guard(
+                cx,
+                outer_arm,
+                if_expr.span,
+                pat_binding,
+                field_name,
+                pat.span,
+                None,
+                &mut app,
+            );
+        }
+    }
+}
+
+fn remove_binding(
+    cx: &LateContext<'_>,
+    pat: &Pat<'_>,
+    span: Span,
+    app: &mut Applicability,
+) -> (Cow<'static, str>, Cow<'static, str>) {
+    let pat_start = snippet_with_applicability(cx, pat.span.with_hi(span.lo()), "<start>", app);
+    let pat_end = snippet_with_applicability(cx, pat.span.with_lo(span.hi()), "<end>", app);
+    (pat_start, pat_end)
+}
+
+fn get_pat_binding(cx: &LateContext<'_>, guard_expr: &Expr<'_>, outer_arm: &Arm<'_>) -> Option<(Span, Option<Ident>)> {
+    if let ExprKind::Path(qpath) = guard_expr.kind
+        && let res = cx.qpath_res(&qpath, guard_expr.hir_id)
+        && let Res::Local(hir_id) = res
+        && local_came_from_and_not_used(cx, &res, outer_arm.pat.hir_id, outer_arm.body)
+    {
+        return cx.tcx.hir().res_span(res).map(|span| {
+            (
+                span,
+                match cx.tcx.hir().get_parent(hir_id) {
+                    Node::PatField(PatField { ident, .. }) => Some(*ident),
+                    _ => None,
+                },
+            )
+        });
+    }
+
+    None
+}
+
+fn get_guard(cx: &LateContext<'_>, guard: Option<Guard<'_>>, app: &mut Applicability) -> Cow<'static, str> {
+    guard.map_or_else(
+        || Cow::Borrowed(""),
+        |guard| {
+            let (prefix, span) = match guard {
+                Guard::If(e) => ("if", e.span),
+                Guard::IfLet(l) => ("if let", l.span),
+            };
+
+            Cow::Owned(format!(
+                " {prefix} {}",
+                snippet_with_applicability(cx, span, "<guard>", app),
+            ))
+        },
+    )
+}
+
+fn local_came_from_and_not_used(cx: &LateContext<'_>, local: &Res, target: HirId, body: &Expr<'_>) -> bool {
+    if let Res::Local(local) = local
+        && cx.tcx.hir().parent_iter(*local).any(|(p, _)| p == target)
```

Comment:
We ought to directly check the `Pat` for the bindings so that we can avoid linting if it's bound on several sides of a `|`

e,g. `A(x) | B(x) | C(x) if matches!(x, something) => { .. }`

https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/struct.Pat.html#method.each_binding is probably the easiest way of doing this